aten/src/ATen/SparseGCSTensorImpl.cpp

```diff
@@ -0,0 +1,130 @@
+  sizes_ = src.sizes();
+}
+  
+void SparseGCSTensorImpl::set_member_tensors_unsafe(const Tensor& pointers, const Tensor& indices,
+                                                      const Tensor& values, const Tensor& reduction,
+                                                    const Scalar& fill_value) {
+  // TODO: perform lots of error checking to check correct type and sizes of inputs. Check
+  // SparseTensorImpl::set_indices_and_values_unsafe() for details
+  TORCH_CHECK(!indices.is_sparse(), "expected indices to be a dense tensor, but got indices of layout ", indices.layout());
+  TORCH_CHECK(!pointers.is_sparse(), "expected pointers to be a dense tensor, but got pointers of layout ", pointers.layout());
+  TORCH_CHECK(!values.is_sparse(), "expected values to be a dense tensor, but got values of layout ", values.layout());
+  TORCH_CHECK(!reduction.is_sparse(), "expected reduction to be a dense tensor, but got reduction of layout ", reduction.layout());
+
+  TORCH_CHECK(values.device().type() == device().type(), "device type of values (", values.device().type(),
+              ") must match device type of device().type()", device().type(), ")");
+  TORCH_CHECK(!indices.is_cuda() || indices.get_device() == values.get_device(), "device of indices (", indices.get_device(),
+              ") must match device of values (", values.get_device(), ")");
+  TORCH_CHECK(!pointers.is_cuda() || pointers.get_device() == values.get_device(), "device of pointers (", pointers.get_device(),
+              ") must match device of values (", values.get_device(), ")");
+
+  TORCH_CHECK(indices.size(0) == values.size(0), "indices and values must have same nnz, but got nnz from indices: ",
+              indices.size(0), ", nnz from values: ", values.size(0));
+
+  TORCH_CHECK((((indices.dim() == pointers.dim()) == values.dim()) == reduction.dim()) == 1,
+              "indices, pointers, values and reduction must have dim=1 but got indices.dim()=",
+              indices.dim(), " pointers.dim()=", pointers.dim(), " values.dim()=", values.dim(), " reduction.dim()=", reduction.dim());
+  
+  pointers_ = pointers;
+  indices_ = indices;
+  values_ = values;
+  reduction_ = reduction;
+  fill_value_ = fill_value;
+
+  AT_ASSERT(device() == values_.device());    
+  AT_ASSERT(indices_.device() == values_.device());
+  AT_ASSERT(reduction_.device() == values_.device());
+
+  auto reduction_accessor = reduction_.accessor<int64_t, 1>();
+
+  rsplit_dim_ = reduction_accessor[reduction_.size(0)-1];
+    
+  dims0_.resize(rsplit_dim_);
+  strides0_.resize(1);
+  
+  dims1_.resize(sizes_.size() - rsplit_dim_);
+  strides1_.resize(1);
+  
+  for (int i = 0; i < rsplit_dim_; ++i) { dims0_[i] = i; }
+  for (int i = 0; i < sizes_.size() - rsplit_dim_; ++i) { dims1_[i] = i + rsplit_dim_; }
+
+  make_strides(0, strides0_, dims0_);
+  make_strides(rsplit_dim_, strides1_, dims1_);
```

Comment:
I guess it's in the roadmap but I just want to notice that the `set` method assumes that the tensor lives on CPU mem. Moreover, I was curious to see the purpose of `strines_` and   and `dims_`   members, it looks like these members are used to convert to a dense tensor. So for CUDA tensors, these part of code and these members should be on GPU too.

Reply:
Thank you for noticing that. I've taken a note of it. I'll look into once we start adding CUDA support. Lets get the implementation logic out of the way first.